### ISSUE TITLE
Skia backend

### ISSUE THREAD

**Issue by RazrFalcon** [RazrFalcon/resvg#44]

**Comment by velyan** [comment]
Hey @RazrFalcon,

I'm gonna start doing some work towards this. 

I'll probably just go by the example of the [cairo backend](https://github.com/RazrFalcon/resvg/tree/master/src/backend_cairo).

I'll add updates here along the way.

**Comment by RazrFalcon** [comment]
I'm all in. Also, I think you should use `new-text-layout` branch as default, since it soon will be merged into the master. So you don't have to worry about text rendering.

I didn't done this task yet, mainly because there are no good skia bindings last time I've checked and I'm not familiar with skia in general.

Feel free to ask any questions.

**Comment by JaFenix** [comment]
Hi, if it's of interest, I am currently working on a Skia backend that is source compatible with resvg v0.7.0.  It is feature complete at this point and I'm focused now on testing/fixing/optimizing.  It uses an older version of Skia from 2017, but with a little effort could be updated to use the latest.   (Our application requires the older Skia version, so I am stuck with that one for now).  I implemented the bindings similar to the Qt bindings, so it does not use rust-skia.

The resvg and Skia forks are located at my [user account](https://github.com/JaFenix).  There are some instructions for building Skia in the skia repo.  When building resvg, the environment variables SKIA_DIR and SKIA_LIB_DIR should be set.  SKIA_DIR must point to the root of the skia directory and SKIA_LIB_DIR should point to ./out/win_x64/release or wherever the skia.dll file is placed.

Cheers,
JaFenix

**Comment by RazrFalcon** [comment]
Hi! Yes, I saw your fork. Great job!

I agree that we should use a custom bindings, like with Qt, because we need just a few features.

I'm only not sure why do you need a skia fork. Can we use an official branch? Is it hard to support an older version, which you need, in the bindings/resvg?

Do you plan to make a pull request? I will try to build your fork tomorrow.

Anyway, building skia on all platforms is the main challenge for me. Can you install (not build, it will take forever) skia on CI?

Also, I'm not sure what is @velyan progress.

PS: If this is not a secret, on what kind of a project do you work on? People are often asking me what is the point in `resvg` and it's kinda hard to explain =)

**Comment by velyan** [comment]
Got other work prioritized over this and haven't made any progress yet. I'm sorry for not updating the issue. 

Great work @JaFenix !

**Comment by RazrFalcon** [comment]
@JaFenix I was able to build it on linux with minimal changes, but rendering doesn't work:

```
Warning (in resvg::backend_skia:159): Failed to create a 194x148 surface.
Error: failed to allocate an image
```

**Comment by JaFenix** [comment]
@RazrFalcon 

Hi!  

> I'm only not sure why do you need a skia fork. Can we use an official branch? Is it hard to support an older version, which you need, in the bindings/resvg?

Yes it is possible to use the official branch @4447b64a directly, but it has a dependency on a version of libjpeg-turbo which can't be resolved without manually editing the DEPS file.  The library appears to have been removed from Google's public repo, so it fails when running the git-sync-deps script.  I found and forked the required version of libjpeg-turbo and then modified the skia DEPS file to reference it.  There is also a memory leak issue, but it usually doesn't appear until after rendering Skia continually for many hours.  Both of these issues have been fixed in the official branch(es) of newer Skia versions.  The only other reason I forked Skia was to provide default args.gn files to make building for win_x64 easier, so someone does not have to create them from scratch.

> Do you plan to make a pull request? 

Yes, we would like to contribute to the project and I can make a pull request.  Which branch in RazrFalcon/resvg should it be directed to?  

> Can you install (not build, it will take forever) skia on CI?

We normally build Skia manually for Windows and Linux as needed (which is a rare occurrence), then commit skia.dll into an application repo, which is built continuously.  However, I may not understand the question correctly.  Is there a CI tool that you use?

> ...on what kind of a project do you work on?

It will be used for HMI/visualization software.  We evaluated all the available SVG solutions we could find but liked resvg the best because it could render all the SVG examples we tried, whereas other libraries failed, were slow, or were out of date with the SVG spec (e.g. Qt, Lead Tools, SharpVectors).  Also, we require a cross-platform library for Window and Linux that can be used natively with C++ or interop with a WPF/.NET app.  The downside was that we did not want to drag Cairo or Qt into our app's process (because we use Skia already), and we were unfamiliar with Rust programming.   However, in the end we concluded that it would still be the best library to use and created the Skia backend for our needs.   

> I was able to build it on linux with minimal changes, but rendering doesn't work

I haven't tried the Skia backend on Linux yet.  I don't know if I'll have the time to troubleshoot as my company focus is on Windows for now.  We've built and run that version of Skia on Debian many times with success, so in theory it should work.  There were issues related to using OpenGL ES surfaces which we had to fix, but I did not commit those into the Skia fork.  However, the backend is creating bitmap/raster surfaces here, so I don't believe GL would be the issue.  Have you been able to build/run it on Windows?  Normally, I have to step into the Skia source to see what's going as it's not well documented.

**Comment by RazrFalcon** [comment]
>it has a dependency on a version of libjpeg-turbo which can't be resolved without manually editing the DEPS file. 

`resvg` doesn't require JPEG support anymore, so we can ignore it.

>Which branch in RazrFalcon/resvg should it be directed to?

master. There were some radical changes a few commits back, but I do not plan any breaking changes any time soon.

>Is there a CI tool that you use?

I want to be able to test the skia backend on the travis CI. But I could not find any Ubuntu PPA's with skia. Travis CI has a 50 min limit and I think that this is not enough to build skia. So we have to provide binaries or manage a PPA. The first one is the easiest.

>I haven't tried the Skia backend on Linux yet.

I will try to fix it. Also, I should try to reproduce it on the latest Skia version.

>Have you been able to build/run it on Windows?

Not yet.

**Comment by JaFenix** [comment]
> There were some radical changes a few commits back...
Yes, I reset back to the v0.7.0 tag because of the FlatRender trait, for which I didn't think I'd have time to refactor.

OK, a pull request is submitted.  Let me know if there's anything else I can do.

**Comment by RazrFalcon** [comment]
Hmm... looks like I cannot add commits to your pull request. I've made a temporary branch on which I will do the rest of the work.

**Comment by RazrFalcon** [comment]
@JaFenix Can you explain a bit how Skia works? I'm not sure about Canvas - Surface relations. In cairo/Qt, we have a Surface and we have to create a Canvas that uses it, to actually draw something on it. But looks like in a Skia, Canvas is a part of Surface? Is it created automatically?

**Comment by RazrFalcon** [comment]
@JaFenix The latest Skia master works fine. Looks like we have to support multiple Skia versions, since 58 -> 76 had some breaking changes.

**Comment by JaFenix** [comment]
Yes, the Canvas is like the drawing tool and context for the surface, so it holds drawing state, like the current matrix and clip region.  The Surface is an opaque region of pixels that could be backed by an in memory bitmap or an Open GL buffer.  By itself, the Surface has no drawing commands, so it gives you the Canvas interface which you use to draw on the surface.  So unlike Qt,  you don't have to create Canvas (Painter), you just ask the surface to give you its Canvas interface so you can draw on the surface.  

Unlike a Painter object, the Canvas doesn't hold color, pen, pattern, or effect state.  The state of these things are stored in a Paint object, and the Paint object is passed as a parameter to all Canvas "draw*" methods.   So basically, you first create a Surface to hold the rasterization results (pixels), acquire a pointer to its Canvas, set the Canvas's matrix and clip stack, create a Paint object and set the colors and styles you want, then call one of the Canvas draw* methods passing in the Paint object accordingly.  You can create as many Paint objects as you want and use them in the draw* methods to the same Canvas.

For the Skia backend, it was not hard to follow the same model of the Qt backend.  Everywhere I saw a qt::Image, I substituted that with a skia::Surface;  and everywhere I saw a qt::Painter being created, I substituted that with a call to Surface::get_canvas().  Instead of setting pen or color state in the Painter instance, I would create Paint objects and set the state there, and then pass it into the draw* method.

This overview on the Skia site might explain [Canvas](https://skia.org/user/api/skcanvas_overview) some more.

**Comment by JaFenix** [comment]
Do you mean that Skia version 76 is working with the skia backend on Linux (i.e. returning a non-null surface pointer)?

**Comment by RazrFalcon** [comment]
Thanks!

Yes, m76 works correctly. And your forks returns a `null`. Not sure why. I've tested it only on Linux.

**Comment by JaFenix** [comment]
It could be that we made some additional fixes for that Skia version to work on Linux a couple years ago.  As long as the Skia API hasn't changed and the bindings in the backend are the same, then it probably isn't necessary to support m58.  I mean, we can substitute in our forked skia.dll when we use it our app, but I don't think resvg needs to build or support our fork for our purposes.

I'm going to do some tests here on Windows to check the binding compatibility.  I just built and ran rendersvg from your latest commit to skia-backend using our skia fork, and that worked fine.  I will try it next using m76.

**Comment by RazrFalcon** [comment]
A few methods are [changed](https://github.com/RazrFalcon/resvg/pull/155/commits/5222f8583b2213d6add4f44a8ba33c04d29a9352).

**Comment by JaFenix** [comment]
Ah, I see that now.  Maybe conditional compilation inside the skia_capi.cpp file, e.g. #define USE_SKIA_M58?

**Comment by RazrFalcon** [comment]
Yes, that was exactly my idea.

**Comment by JaFenix** [comment]
I ran rendersvg against the latest commit (7c9a8d05) and Skia m76 on Windows and everything looks good so far.

**Comment by RazrFalcon** [comment]
Yes. I'm fixing an image rendering right now (which I broke during porting to FlatRender) and that's it.

**Comment by RazrFalcon** [comment]
@velyan Skia backend has been merged. If you still plan to use it, fill free to share your experience.

**Comment by velyan** [comment]
Thanks @RazrFalcon @JaFenix 🙌

**Comment by upsuper** [comment]
> I agree that we should use a custom bindings, like with Qt, because we need just a few features.

I respectfully disagree.

The convention in Rust ecosystem of handling binding to a foreign language library is to have all crates depend on a single binding crate (usually with name ending with `-sys`), which is created in an uncontroversial way (just plain binding over all export functions and types without any abstraction).

The problem having multiple binding crates is that, it would make it very hard for downstream consumers to combine crates with different bindings.

In this case, if someone wants to use this skia backend, very likely they would want to combine it with rust-skia, which would not work properly with a custom binding.

The main difficulty comes from linking. C and C++ are *not* designed to allowing multiple build of the same library being linked in a single binary. It may cause link issue, or even if it's not, there is no guarantee that everything would work as expected.

`skia-bindings` seems to be the one binding crate (although not named with `-sys`) that exposes the raw bindings, and all crates linking to skia should ideally just depend on that crate rather than creating their own bindings.

**Comment by RazrFalcon** [comment]
@upsuper 

There are no good bindings for Skia or Qt. When they became available I will port `resvg` to them.

Also, `skia-bindings` are not on crates.io.

**Comment by upsuper** [comment]
@RazrFalcon Oh, okay, that makes sense then. I didn't realize that `skia-bindings` is not on crates.io. I guess it would be great if they can publish it.

**Comment by RazrFalcon** [comment]
It also should support older Skia versions, because this is required by @JaFenix.

Everything is not that easy when you start to dig in.

**Comment by JaFenix** [comment]
> In this case, if someone wants to use this skia backend, very likely they would want to combine it with rust-skia, which would not work properly if resvg uses a custom binding.

@upsuper Consider that resvg is also a part of the C/C++ ecosystem too, because it has the C API.   Our C++ app already depends on the Skia binary, so the Rust bindings just need to work with the skia.dll that we build and load into our app's process.  It's similar to using resvg with the Qt bindings in a Qt app.   I see the concern over conflicts in the Rust ecosystem, but the C/C++ usage scenarios might be a little different here.  The current bindings work quite well for C/C++ apps because they're slim and don't do more than is needed.

@RazrFalcon We will likely update to a new version of Skia (> m58) at some point in the future.  I don't think resvg should try to support the older m58 version forever.  Moving forward to the latest version makes more sense.  However, Skia's interfaces/API sometimes change between milestone releases, so I think there will always be a need for Skia bindings that work with more than one Skia milestone.

It will not matter to me what binding library is used, but I prefer a minimal set of bindings (like resvg currently has) to a larger set if it means more efficiency inside a C++/Skia app.  Resvg only uses a small set of Skia features, whereas rust-skia includes everything to run Skia in a standalone Rust app.  It would be a shame to use a set of bindings that bloated the code or made it more complicated than necessary.  Skia is a large library and there may not be a good "one size fits all" solution for creating a binding library, especially where two ecosystems are involved.

**Comment by RazrFalcon** [comment]
@JaFenix I agree. Currently, it's not that hard to support older Skia versions, so it's not a problem.
